### ISSUE TITLE
Wrong Typescript types `Ref`

### ISSUE THREAD

**Issue by susnux** [vueuse/vueuse#4343]
### Describe the bug

The vueuse types for `Ref` have two template parameters, but the `Ref` type in Vue2 (at least) only has one.

from vueuse `index.d.mts`:

```ts
declare function useElementSize(target: MaybeComputedElementRef, initialSize?: ElementSize, options?: UseResizeObserverOptions): {
    width: vue.Ref<number, number>;
    height: vue.Ref<number, number>;
    stop: () => void;
};
```

As you can see `width` has type `vue.Ref<number, number>` which is invalid and cause the type `any` to be inferred instead of `Ref<number>`.

### Reproduction

See your source code

### System Info

```Shell
System:
    OS: Linux 6.11 openSUSE Tumbleweed 20241112
    CPU: (16) x64 AMD Ryzen 7 8845HS w/ Radeon 780M Graphics
    Memory: 17.66 GB / 27.21 GB
    Container: Yes
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 22.10.0 - /usr/bin/node
    npm: 10.9.0 - /usr/bin/npm
  npmPackages:
    @vueuse/components: ^11.2.0 => 11.2.0 
    @vueuse/core: ^11.2.0 => 11.2.0 
    @vueuse/integrations: ^11.2.0 => 11.2.0 
    vue: ^2.7.16 => 2.7.16
```

### Used Package Manager

npm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vueuse/vueuse/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vueuse/vueuse/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vueuse.org/guide).
- [X] Check that there isn't [already an issue](https://github.com/vueuse/vueuse/issues) that reports the same bug to avoid creating a duplicate.
- [X] Make sure this is a VueUse issue and not a framework-specific issue. For example, if it's a Vue SFC related bug, it should likely be reported to https://github.com/vuejs/core instead.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vueuse/vueuse/discussions).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by kubajmarek** [comment]
I'm using Vue3 and I started to have the same problem with `useMediaQuery` when attempting to go from 11.0.3 to 11.2.0.

**Comment by susnux** [comment]
Sure I expected that vue2 / vue-demi support will be dropped some when, but the issue seems to be unrelated to that as @kubajmarek apparently has the same issue with Vue3.

I also do not see why `Ref` with two template parameters is used here?

**Comment by OrbisK** [comment]
> Not exactly the same issue as [@susnux](https://github.com/susnux), but typechecks are failing when I updated from `v11.3.0` to `v12.4.0`.
> 
> import { useRouteQuery } from '@vueuse/router'
> 
> const foo = useRouteQuery('foo', 'foo', {
>   mode: 'push',
> })
> 
> <template v-if="foo === 'foo'">
>     Some text
> </template>
> Typescript complains about this:
> 
> > error TS2367: This comparison appears to be unintentional because the types 'Ref<string, string>' and 'string' have no overlap.
> 
> This still works because `Vue` conveniently unwraps the Ref when used in a template but `TypeScript` does not recognize this.
> 
> Do we have a workaround for this? [@antfu](https://github.com/antfu)

can't reproduce this:

```ts
  it('should work with strings', () => {
    let route = getRoute({
      search: 'vue3',
    })
    const router = { push: (r: any) => route = r } as any
    const query = useRouteQuery('foo', 'foo', {
      route,
      router,
      mode: 'push',
    })

    expect(query.value).toBe('foo')
    expect(query.value === 'foo').toBe(true)
  })
```
types are fine for me.

Can you please share a minimal reproduction?

**Comment by antfu** [comment]
We no longer support Vue 2 in v12, consider this issue is resolved.

**Comment by KazariEX** [comment]
> Not exactly the same issue as [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/1855448?v=4&amp;size=80">@susnux](https://github.com/susnux?rgh-link-date=2025-01-21T15%3A59%3A50.000Z), but typechecks are failing when I updated from `v11.3.0` to `v12.4.0`.
> 
> import { useRouteQuery } from '@vueuse/router'
> 
> const foo = useRouteQuery('foo', 'foo', {
>   mode: 'push',
> })
> 
> <template v-if="foo === 'foo'">
>     Some text
> </template>
> Typescript complains about this:
> 
> > error TS2367: This comparison appears to be unintentional because the types 'Ref<string, string>' and 'string' have no overlap.
> 
> This still works because `Vue` conveniently unwraps the Ref when used in a template but `TypeScript` does not recognize this.
> 
> Do we have a workaround for this? [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/11247099?u=746eca82e375631967235dd01261428762bbcbf9&amp;v=4&amp;size=80">@antfu](https://github.com/antfu?rgh-link-date=2025-01-21T15%3A59%3A50.000Z)

![Image](https://github.com/user-attachments/assets/8a38e940-6edd-4b4f-aaa3-0cd1d20716e6)

**Comment by FaitAccompli** [comment]
@KazariEX thanks, but I'm not getting that error, rather I'm getting this:
> This comparison appears to be unintentional because the types 'Ref<string, string>' and 'string' have no overlap.ts-plugin(2367)
Whenever I use the `Ref` in a `v-if` for example. The strange thing is, I'm trying to reproduce it [here](https://stackblitz.com/edit/vitejs-vite-xkyp9g3s?file=src%2FApp.vue) to no avail.

**Comment by ilyaliao** [comment]
> [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/85992002?u=df5f2240ec250b848e2fc0c3a4c64bb7a27f9ab0&amp;v=4&amp;size=80">@KazariEX](https://github.com/KazariEX) thanks, but I'm not getting that error, rather I'm getting this:
> 
> > This comparison appears to be unintentional because the types 'Ref<string, string>' and 'string' have no overlap.ts-plugin(2367)
> 
> Whenever I use the `Ref` in a `v-if` for example. The strange thing is, I'm trying to reproduce it [here](https://stackblitz.com/edit/vitejs-vite-xkyp9g3s?file=src%2FApp.vue) to no avail.

Could you let me check your Vue and TypeScript versions? I'm trying to reproduce.

**Comment by FaitAccompli** [comment]
Sure @ilyaliao! 
```
"vue": "3.5.13"
"typescript": "5.7.3"
"vue-tsc": "2.2.0"
"@vueuse/core": "12.5.0"
"@vueuse/router": "12.5.0"
```

**Comment by ilyaliao** [comment]
Yes, after upgrading to `vue@3.5.13`, the error disappeared @FaitAccompli. The issue can be reproduced in `vue@3.5.12`.

I think this is another new issue, related to: #3240. It seems that upgrading the version solves it.

**Comment by FaitAccompli** [comment]
@ilyaliao Thanks for confirming. For those experiencing the same issue. Here's a summary: 

This works fine ... 
```
"vue": "3.5.12"
"@vueuse/core": "11.3.0"
"@vueuse/router": "11.3.0"
```
But this doesn't 
```
"vue": "3.5.12"
"@vueuse/core": "12.5.0"
"@vueuse/router": "12.5.0"
```
This fixes it
```
"vue": "3.5.13"
"@vueuse/core": "12.5.0"
"@vueuse/router": "12.5.0"
```
